mypy/plugins/ctypes.py

```diff
@@ -0,0 +1,185 @@
+"""Plugin to provide accurate types for some parts of the ctypes module."""
+
+from typing import Optional
+
+# Fully qualified instead of "from mypy.plugin import ..." to avoid circular import problems.
+import mypy.plugin
+from mypy.maptype import map_instance_to_supertype
+from mypy.subtypes import is_subtype
+from mypy.types import CallableType, Instance, Type, UnionType, union_items
+
+
+def _find_simplecdata_base_arg(tp: Instance, api: 'mypy.plugin.CheckerPluginInterface'
+                               ) -> Optional[Type]:
+    """Try to find a parametrized _SimpleCData in tp's bases and return its single type argument.
+
+    None is returned if _SimpleCData appears nowhere in tp's (direct or indirect) bases.
+    """
+    if tp.type.has_base('ctypes._SimpleCData'):
+        simplecdata_base = map_instance_to_supertype(tp,
+            api.named_generic_type('ctypes._SimpleCData', []).type)
+        assert len(simplecdata_base.args) == 1, '_SimpleCData takes exactly one type argument'
+        return simplecdata_base.args[0]
+    return None
+
+
+def _autoconvertible_to_cdata(tp: Type, api: 'mypy.plugin.CheckerPluginInterface') -> Type:
+    """Get a type that is compatible with all types that can be implicitly converted to the given
+    CData type.
+
+    Examples:
+    * c_int -> Union[c_int, int]
+    * c_char_p -> Union[c_char_p, bytes, int, NoneType]
+    * MyStructure -> MyStructure
+    """
+    # Every type can be converted from itself (obviously).
+    allowed_types = [tp]
+    if isinstance(tp, Instance):
```

Comment:
We should, but this isn't as straightforward as with the attribute hooks. The type returned from `_autoconvertible_to_cdata` is not used as a return type, but an argument type, for example in `Array.__setitem__`. This makes the rules for union types a little more complicated here.

For example, if `arr` has the type `Array[Union[c_int, c_uint]]`, `arr[i] = x` is only valid if `x` can be converted to both `c_int` *and* `c_uint`, i. e. the type of `x` should be the intersection of `_autoconvertible_to_cdata(c_int)` and `_autoconvertible_to_cdata(c_uint)`.

As far as I know, mypy doesn't support intersection types, and I can't do an easy intersection calculation using sets, since `Type` objects are not hashable. So I guess I'd have to write a manual intersection algorithm based on `is_subtype`?

Reply:
Mypy types are hashable (we have a subtype cache for example that uses this), but I am not sure this is what you need. Also I am not sure we need an intersection here, it is indeed safe, but I am afraid it may cause false positives. I am fine with a less strict check (especially if it is easy to implement). IOW I don't worry about false negatives because we don't allow unions here, I worry about false positives.

Reply:
> Mypy types are hashable

Ah, my mistake - I only looked at the `Type` base class and saw that it had no `__hash__`, but almost all of the concrete type classes define `__hash__`.

> IOW I don't worry about false negatives because we don't allow unions here, I worry about false positives.

In that case I can simply add another "union-loop" thing. But tbh even if this is changed later, I don't think it will give many unexpected errors - an `Array[Union[...]]` type can only come from an explicit annotation, in which case the user perhaps wants this kind of strict checking.